sklearn/metrics/_scorer.py

```diff
@@ -636,39 +507,72 @@ def _check_multimetric_scoring(estimator, scoring):
         }
     else:
         raise ValueError(err_msg_generic)
-    return scorers
+
+    # To be cache friendly, we reduce the number of response methods to a single one
+    # based on what the estimator supports.
+    final_scorers = {}
+    for name, scorer in scorers.items():
+        if not isinstance(scorer, _BaseScorer) or isinstance(
+            scorer._response_method, str
+        ):
+            final_scorers[name] = scorer
+        else:
+            # Sanity check that the scorer has the same init arguments as `_BaseScorer`.
+            # Otherwise, we are going to miss some arguments when creating the object.
+            base_scorer_params = set(signature(_BaseScorer).parameters.keys())
+            scorer_params = set(signature(scorer.__class__).parameters.keys())
+            if not (base_scorer_params == scorer_params):
+                final_scorers[name] = scorer
+            else:
+                final_response_method = _check_response_method(
+                    estimator, scorer._response_method
+                )
+                new_scorer = scorer.__class__(
+                    scorer._score_func,
+                    scorer._sign,
+                    scorer._kwargs,
+                    response_method=final_response_method.__name__,
+                )
+                if hasattr(scorer, "_metadata_request"):
+                    new_scorer._metadata_request = scorer._metadata_request
```

Comment:
This is indeed the magic part that recreate a scorer from a previous one and just change the `response_method` to help for caching.